### DISCUSSION TITLE
dont show suggestions longer than X letters?

### DISCUSSION THREAD

**Discussion by Astrantia** [hrsh7th/nvim-cmp#609] [Q&A]
Hi, I would like to hide my suggestions that are longer than 10 letters. Usually I never accept verylongsuggestionslikethis. How can I do it?

**Comment by dmitmel** [comment]
This is not _exactly_ what you want, but you may find this solution useful: you can use the `formatting.format` setting to truncate long items to a desired length, like this:

```lua
local ELLIPSIS_CHAR = '…'
local MAX_LABEL_WIDTH = 20

cmp.setup({
  formatting = {
    format = function(entry, vim_item)
      local label = vim_item.abbr
      local truncated_label = vim.fn.strcharpart(label, 0, MAX_LABEL_WIDTH)
      if truncated_label ~= label then
        vim_item.abbr = truncated_label .. ELLIPSIS_CHAR
      end
    end,
  },
})

**Reply by krishnakumarg1984** [reply]
Is there something equivalent to enforcing a minimum width of the `cmp` autocompletion window?

**Reply by krishnakumarg1984** [reply]
Or even better, how to keep cmp's pop-up menu width **fixed**? It is annoying if the completion menu's width keeps on changing while typing. It's a very jagged experience

**Reply by zhengjingxianghuo** [reply]
How do I limit the height of the completion menu since it almost fills my screen?

**Comment by kkharji** [comment]
Based on @dmitmel  but it tries to keep the width consistent (note: scroll adds extra space)

```lua
local ELLIPSIS_CHAR = '…'
local MAX_LABEL_WIDTH = 25
local MAX_KIND_WIDTH = 14

local get_ws = function (max, len)
  return (" "):rep(max - len)
end

local format = function(_, item)
  local content = item.abbr
  -- local kind_symbol = symbols[item.kind]
  -- item.kind = kind_symbol .. get_ws(MAX_KIND_WIDTH, #kind_symbol)

  if #content > MAX_LABEL_WIDTH then
    item.abbr = vim.fn.strcharpart(content, 0, MAX_LABEL_WIDTH) .. ELLIPSIS_CHAR
  else
    item.abbr = content .. get_ws(MAX_LABEL_WIDTH, #content)
  end

  return item
end

cmp.setup({
  formatting = {
    format = format,
  },
})
```

cc @krishnakumarg1984

**Comment by Roundlay** [comment]
Here's a version of the completion entry truncation function that adapts the truncation and padding of completion entries based on Neovim's window width, or uses a fixed width for the completion menu, depending on the provided configuration, as per @krishnakumarg1984's comment.

When `fixed_width` isn't false, it fixes the completion window width and adjusts the completion entries accordingly. If the `fixed_width` parameter isn't provided or is false, the function will use a dynamic width based on a percentage of the Neovim window width.

Here's what it looks like inside the formatting block in my `cmp` setup function:

```lua
formatting = {
    fields = { "abbr", "menu", "kind" },
    format = function(entry, item)
        -- Define menu shorthand for different completion sources.
        local menu_icon = {
            nvim_lsp = "NLSP",
            nvim_lua = "NLUA",
            luasnip  = "LSNP",
            buffer   = "BUFF",
            path     = "PATH",
        }
        -- Set the menu "icon" to the shorthand for each completion source.
        item.menu = menu_icon[entry.source.name]

        -- Set the fixed width of the completion menu to 60 characters.
        -- fixed_width = 20

        -- Set 'fixed_width' to false if not provided.
        fixed_width = fixed_width or false

        -- Get the completion entry text shown in the completion window.
        local content = item.abbr

        -- Set the fixed completion window width.
        if fixed_width then
            vim.o.pumwidth = fixed_width
        end

        -- Get the width of the current window.
        local win_width = vim.api.nvim_win_get_width(0)

        -- Set the max content width based on either: 'fixed_width'
        -- or a percentage of the window width, in this case 20%.
        -- We subtract 10 from 'fixed_width' to leave room for 'kind' fields.
        local max_content_width = fixed_width and fixed_width - 10 or math.floor(win_width * 0.2)

        -- Truncate the completion entry text if it's longer than the
        -- max content width. We subtract 3 from the max content width
        -- to account for the "..." that will be appended to it.
        if #content > max_content_width then
            item.abbr = vim.fn.strcharpart(content, 0, max_content_width - 3) .. "..."
        else
            item.abbr = content .. (" "):rep(max_content_width - #content)
        end
        return item
    end,
},
```

Fixed width menu example:

![Screenshot (49)](https://user-images.githubusercontent.com/4133752/234488420-7d73b04d-3904-4504-9bf2-8826ca687755.png)
![Screenshot (50)](https://user-images.githubusercontent.com/4133752/234488434-81510a08-cbf8-402a-950b-490d2fbefe71.png)

Dynamic width menu example:

![Screenshot (52)](https://user-images.githubusercontent.com/4133752/234488457-4d2456c8-907f-40d5-8ab8-5d0cb2449489.png)
![Screenshot (53)](https://user-images.githubusercontent.com/4133752/234488469-2a1411dd-e16b-46d2-9eca-07c7be748243.png)

**Comment by kleinepls** [comment]
is it possible to do this for the documentation window as well?

**Reply by aguydan** [reply]
Actually it is! I know I'm a month late but I just found a way to adjust the size of the docs window.

```lua
cmp.setup({
    window = {
       documentation = {
          max_height = 30,
          max_width = 20,
          -- other settings
       }
    }
})
```

**Reply by bora89** [reply]
<img width="3022" height="1514" alt="tmux-5xeVd11c" src="https://github.com/user-attachments/assets/3bc4172d-b843-4841-b294-be2df4308a48" />
Did not work for some reason - still huge screen-wide
